File: Matrix.java
import java.util.Random;
import java.util.function.Function;

public class Matrix {

    private float[][] values;
    public int m, n;

    //init matrix size m*n with all entries 0
    public Matrix(int m, int n){
        values = new float[m][n];
        this.m = m;
        this.n = n;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                put(i, j, 0);
            }
        }
    }

    //substiture entry i,j with value
    public void put(int i, int j, float value){
        values[i][j] = value;
    }

    public void toVectorColumn(float[] values){
        m = 1;
        n = values.length;
        this.values = new float[][]{values};
        this.values = Matrix.transpose(this).toArray();
        m = values.length;
        n = 1;
    }

    //get entry at i,j
    public float get(int i, int j) {
        return values[i][j];
    }

    //randomize every entry of matrix
    public void randomize(int lowerbound, int upperbound){
        Random random = new Random();
        random.setSeed(90);
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                //values[i][j] = lowerbound + random.nextFloat() * (upperbound - lowerbound);
                values[i][j] = random.nextInt(10)-5;
            }
        }
    }

    //get the matrix as array
    public float[][] toArray(){
        return values;
    }

    public float[] getColumn(int col){
        float[] column = new float[m];
        for(int i=0; i<m; i++){
            column[i] = get(i, col);
        }
        return column;
    }

    public float[] getRow(int row){
        return values[row];
    }

    public static Matrix multiply(Matrix matrix1, Matrix matrix2) {
        Matrix result = new Matrix(matrix1.m, matrix2.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix2.n; j++){
                float sum = 0;
                for(int k=0; k<matrix1.n; k++){
                    sum += matrix1.get(i, k)*matrix2.get(k, j);
                }
                result.put(i, j, sum);
            }
        }
        return result;
    }

    static public Matrix transpose(Matrix matrix1){
        Matrix result = new Matrix(matrix1.n, matrix1.m);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(j, i, matrix1.get(i, j));
            }
        }
        return result;
    }

    //apply function to every entry of matrix
    public static Matrix map(Matrix matrix1, Function<Float, Float> function){
        Matrix result = new Matrix(matrix1.m, matrix1.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(i, j, function.apply(matrix1.get(i, j)));
            }
        }
        return result;
    }

    public static Matrix add(Matrix matrix1, Matrix matrix2) {
        Matrix result = new Matrix(matrix1.m, matrix1.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(i, j, matrix1.get(i, j)+matrix2.get(i, j));
            }
        }
        return result;
    }

    public static Matrix subtract(Matrix matrix1, Matrix matrix2) {
        Matrix result = new Matrix(matrix1.m, matrix1.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(i, j, matrix1.get(i, j)-matrix2.get(i, j));
            }
        }
        return result;
    }

    //elementwise multiplication of two matrices
    public static Matrix hadamard_product(Matrix matrix1, Matrix matrix2){
        Matrix result = new Matrix(matrix1.m, matrix1.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(i, j, matrix1.get(i, j)*matrix2.get(i, j));
            }
        }
        return result;
    }

    public static Matrix scale(Matrix matrix1, float scale){
        Matrix result = new Matrix(matrix1.m, matrix1.n);
        for(int i=0; i<matrix1.m; i++){
            for(int j=0; j<matrix1.n; j++){
                result.put(i, j, matrix1.get(i, j)*scale);
            }
        }
        return result;
    }

    public float sumAbs(){
        float sum = 0;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                sum += Math.abs(get(i, j));
            }
        }
        return sum;
    }

    public String toString(){
        String s = "";
        for(float[] row: values){
            for(float entry: row){
                s += String.format("%.2f ", entry);
            }
            s += "\n";
        }
        return s;
    }

}

File: JSLibrary/Matrix.js
class Matrix{

    //create empty matrix with dim m*n
    constructor(m, n){
        this.m = m;
        this.n = n;
        this.values = Array(m).fill().map(() => Array(n).fill(0));
    }

    fromArray(array){
        this.m = array.length;
        this.n = 1;
        this.values = Array(this.m).fill().map(() => Array(this.n).fill(0));
        this.map((x,i,j) => array[i])
    }

    toArray() {
        let arr = [];
        for (let i = 0; i < this.m; i++) {
            for (let j = 0; j < this.n; j++) {
                arr.push(this.values[i][j]);
            }
        }
        return arr;
    }

    copy(){
        let matrix1 = new Matrix(this.m, this.n);
        for (let i = 0; i < this.m; i++) {
            for (let j = 0; j < this.n; j++) {
                matrix1.values[i][j] = this.values[i][j];
            }
        }
        return matrix1;
    }

    put(i, j, value){
        this.values[i][j] = value;
    }

    get(i, j){
        return this.values[i][j];
    }

    randomize(lowerbound, upperbound){
        this.map(x => random(lowerbound, upperbound))
    }

    absvalue(){
        let sum = 0;
        for(let i=0; i<this.m; i++){
            for(let j=0; j<this.n; j++){
                sum += abs(this.get(i, j));
            }
        }
        return sum;
    }

    //apply function to every entry
    map(funct){
        for(let i=0; i<this.m; i++){
            for(let j=0; j<this.n; j++){
                this.put(i, j, funct(this.get(i, j), i, j))
            }
        }
    }

    static map(funct, matrix1){
        let matrix2 = matrix1.copy()
        matrix2.map(funct);
        return matrix2;
    }

    add(matrix1){
        this.map((x, i, j) => x + matrix1.values[i][j]);
    }

    static add(matrix1, matrix2){
        let matrix3 = matrix1.copy();
        matrix3.add(matrix2);
        return matrix3;
    }

    sub(matrix1){
        this.map((x, i, j) => x - matrix1.values[i][j]);
    }

    static sub(matrix1, matrix2){
        let matrix3 = matrix1.copy();
        matrix3.sub(matrix2);
        return matrix3;
    }

    transpose(){
        let matrix1 = new Matrix(this.n, this.m);
        matrix1.map((x, i, j) => this.values[j][i])
        this.values = matrix1.values;
        this.m = matrix1.m;
        this.n = matrix1.n;
    }

    static transpose(matrix1){
        let matrix2 = matrix1.copy();
        matrix2.transpose();
        return matrix2;
    }

    scale(t){
        this.map(x => t*x);
    }

    static scale(matrix1, t){
        let matrix2 = matrix1.copy();
        matrix2.scale(t);
        return matrix2;
    }

    hadamard(matrix1){
        this.map((x, i, j) => x * matrix1.values[i][j]);
    }

    static hadamard(matrix1, matrix2){
        let matrix3 = matrix1.copy();
        matrix3.hadamard(matrix2);
        return matrix3;
    }

    multiply(matrix1){
        let new_values = [];
        for(let i=0; i<this.m; i++){
            let row = [];
            for(let j=0; j<matrix1.n; j++){
                let sum = 0;
                for(let k=0; k<this.n; k++){
                    sum += this.get(i, k) * matrix1.get(k, j);
                }
                row.push(sum);
            }
            new_values.push(row);
        }
        this.values = new_values;
        this.n = matrix1.n;
    }

    static multiply(matrix1, matrix2){
        let matrix3 = matrix1.copy();
        matrix3.multiply(matrix2);
        return matrix3;
    }

}